Add Kubernetes monitoring docs (#151)
elastic/observability-docs: docs/en/observability/monitor-k8s/monitor-k8s-logs.asciidoc

------------------------------------------------

Log events should start flowing to Elasticsearch.

** Red Hat OpenShift configuration

If you are using Red Hat OpenShift, you need to specify additional settings in
the manifest file and enable the container to run as privileged.

. Modify the `DaemonSet` container spec in the manifest file:
+
[source,yaml]
-----
  securityContext:
    runAsUser: 0
    privileged: true
-----

. Grant the `filebeat` service account access to the privileged SCC:
+
[source,shell]
-----
oc adm policy add-scc-to-user privileged system:serviceaccount:kube-system:filebeat
-----
+
This command enables the container to be privileged as an administrator for
OpenShift.

. Override the default node selector for the `kube-system` namespace (or your
custom namespace) to allow for scheduling on any node:
+
[source,shell]
----
oc patch namespace kube-system -p \
'{"metadata": {"annotations": {"openshift.io/node-selector": ""}}}'
----
+
This command sets the node selector for the project to an empty string. If you
don't run this command, the default node selector will skip master nodes.

[discrete]
=== View logs in {kib}

> ==== Using the Logs app in Kibana

Author: ```suggestion

[discrete]
==== Using the Logs app in Kibana
```

TODO: Describe how to view logs in Kibana. Show how to use the Logs app and how to
set up and view pre-built dashboards and visualizations.

The https://www.elastic.co/log-monitoring[Logs app] in Kibana allows you to search, filter and tail all the logs
collected into Elastic Stack. Instead of having to ssh into different servers, having to cd into the directory
and tail individual files, all the logs are available in one tool under Logs app.

* Check out filtering logs using the keyword or plain text search.
* You can move back and forth in time using the time picker or the timeline view on the side.
* If you just want to watch the logs update in front of you tail -f style, click on the Streaming button
and use highlighting to accentuate that important bit of the info you are waiting to see.

TODO: Add screenshot here?

> ==== Out-of-the-box Kibana visualisations

Author: ```suggestion

[discrete]
==== Out-of-the-box Kibana visualisations
```

[discrete]
[[monitor-kubernetes-logs]]
== Part 1: Monitor logs

Collecting and analyzing logs of both Kubernetes Core components and various
applications running on top of Kubernetes is a powerful tool for Kubernetes
observability. Containers running within Kubernetes pods produce logs as stdout
or stderr. These logs are written to a location known to Kubelet.

To collect pod logs, all you need is {filebeat} running as a DaemonSet
in your Kubernetes cluster. You configure {filebeat} to communicate with the
Kubernetes API server, get the list of pods running on the current host, and
collect the logs the pods are producing. Those logs are annotated with all the
relevant Kubernetes metadata, such as pod ID, container name, container labels
and annotations, and so on.

[discrete]
=== Deploy {filebeat} to collect logs

To start collecting logs, deploy and run an instance of {filebeat} on each
Kubernetes host. {filebeat} communicates with the Kubernetes API server to
retrieve information about the pods running on the host, all the metadata

> annotations, and the location of the log files.

Reviewer: Log location is not know to Filebeat via k8s API. It is defined in the container input section, for instance at https://github.com/elastic/beats/blob/master/deploy/kubernetes/filebeat-kubernetes.yaml#L14

Author: Right, I must have done a bad edit at some point. :-) I think I meant that it uses annotations to decide which logging modules to apply. But looking a this now, I'm not sure the detail belongs here.

To deploy {filebeat} to your Kubernetes cluster:

[discrete]
==== Step 1: Download the {filebeat} deployment manifest

To make deployment easier, Elastic provides a YAML file that defines all the
required deployment settings. In many cases, you can change the connection
details and deploy with default settings to get started quickly.

["source", "sh", subs="attributes"]
----
curl -L -O https://raw.githubusercontent.com/elastic/beats/{branch}/deploy/kubernetes/filebeat-kubernetes.yaml
----

[discrete]
==== Step 2: Set the connection information for {es}

> By default, {filebeat} sends events to an existing {es} deployment, if present.

Reviewer 2: I noticed the same thing in the Metrics section.

To specify a different destination, change the following settings in the
`filebeat-kubernetes.yaml` file:

[source,yaml]
----
env:
- name: ELASTICSEARCH_HOST
  value: elasticsearch
- name: ELASTICSEARCH_PORT
  value: "9200"
- name: ELASTICSEARCH_USERNAME
  value: elastic
- name: ELASTICSEARCH_PASSWORD
  value: changeme
- name: ELASTIC_CLOUD_ID <1>
  value:

> - name: ELASTIC_CLOUD_AUTH <1>

Author: good idea!

> - name: ELASTIC_CLOUD_AUTH <1>

Author: @DanRoscigno I want to get this review closed out ASAP. Can I mark this as "resolved" with the assumption that you've added this to your todo list?

  value:
----
<1> Use the cloud settings if you're sending data to {ess} on {ecloud}.

To avoid exposing sensitive data, you can base64 encode the string, then store it
in a Kubernetes secret. For example:

["source", "sh", subs="attributes"]
------------------------------------------------
$ echo -n 'changeme' | base64
Y2hhbmdlbWU=
$ kubectl create secret generic es-secret --from-literal=password='Y2hhbmdlbWU=' --namespace=kube-system <1>
------------------------------------------------
<1> Create the secret in the namespace where you will deploy {filebeat}.

To use the secret, change the `env` setting in the manifest file:

[source,yaml]
------------------------------------------------
env:

> - name: ELASTICSEARCH_PASSWORD

Author: Good point! I will add a pointer.

[role="screenshot"]
image::images/log-stream.png[Logs app streaming messages collected by {filebeat}]

Explore the Logs app:

* Enter a keyword or text string in the search field to filter logs. 
* Use the time picker or timeline view on the side to move forward and back in
time.
* Click **Stream live** to watch the logs update in front of you `tail -f`
style.
* Place your cursor over a log message to highlight it, then use the context
menu to view details or view the log message in context. 

[discrete]
==== Out-of-the-box {kib} dashboards

{filebeat} ships with a variety of pre-built {kib} dashboards that you can
use to visualize logs from Kubernetese Core components and applications running
on top of Kubernetes. If these dashboards are not already loaded into {kib}, you
must run the {filebeat} setup job. 

TIP: To run the setup job, install {filebeat} on any system that can connect to
the {stack}, enable the modules for the datasets you want to monitor, then run
the `setup` command. To learn how, see the
{filebeat-ref}/filebeat-installation-configuration.html[{filebeat} quick start].

//TODO: We might want to provide these steps inline (maybe in a collapsed)
//section.

> After the dashboards are loaded, navigate to dashboards for the applications

Author: I don't like this change. Fixing it in the next commit I push.

> After the dashboards are loaded, navigate to dashboards for the applications
> you're running. For example, navigate to the dashboards for MySQL or NGINX.

Author: ```suggestion
After loading the dashboards, navigate to **{kib} > Dashboards**
and search for the services you want to monitor, like MySQL or NGINX.
```

> you can navigate to the {filebeat} dashboards for MySQL and NGINX.

Author: Yes, I think we should. If we decide to use a sample app, we can add it in a later iteration.

available in one tool under the Logs app.

[role="screenshot"]
image::images/log-stream.png[Logs app streaming messages collected by {filebeat}]

Explore the Logs app:

* Enter a keyword or text string in the search field to filter logs. 
* Use the time picker or timeline view on the side to move forward and back in
time.
* Click **Stream live** to watch the logs update in front of you `tail -f`
style.
* Place your cursor over a log message to highlight it, then use the context
menu to view details or view the log message in context. 

[discrete]
==== Out-of-the-box {kib} dashboards

{filebeat} ships with a variety of pre-built {kib} dashboards that you can
use to visualize logs from Kubernetes Core components and applications running
on top of Kubernetes. If these dashboards are not already loaded into {kib}, you
must run the {filebeat} setup job. 

TIP: To run the setup job, install {filebeat} on any system that can connect to
the {stack}, enable the modules for the datasets you want to monitor, then run
the `setup` command. To learn how, see the
{filebeat-ref}/filebeat-installation-configuration.html[{filebeat} quick start].

> //TODO: We might want to provide these steps inline (maybe in a collapsed)

Author: ```suggestion
```

> //TODO: We might want to provide these steps inline (maybe in a collapsed)
> //section.

Author: ```suggestion
```
